packages/next-codemod/bin/upgrade.ts

```diff
@@ -74,14 +74,44 @@ export async function runUpgrade(
     return
   }
 
+  const installedReactVersion = getInstalledReactVersion()
+  console.log(`Current React version: v${installedReactVersion}`)
+  let shouldStayOnReact18 = false
+  if (
+    // From release v14.3.0-canary.45, Next.js expects the React version to be 19.0.0-beta.0
+    // If the user is on a version higher than this but is still on React 18, we ask them
+    // if they still want to stay on React 18 after the upgrade.
+    // IF THE USER USES APP ROUTER, we expect them to upgrade React to > 19.0.0-beta.0,
+    // we should only let the user stay on React 18 if they are using pure Pages Router.
+    // x-ref(PR): https://github.com/vercel/next.js/pull/65058
+    // x-ref(release): https://github.com/vercel/next.js/releases/tag/v14.3.0-canary.45
+    compareVersions(installedNextVersion, '14.3.0-canary.45') >= 0 &&
+    installedReactVersion.startsWith('18')
+  ) {
+    const shouldStayOnReact18Res = await prompts(
+      {
+        type: 'confirm',
+        name: 'shouldStayOnReact18',
+        message: `Are you using ${pc.underline('Pages Router only')} (no App Router) and want to stay on React 18?`,
```

Comment:
“Are you using only the Pages Router (without the App Router) and prefer to stay on React 18?”